### ISSUE TITLE
Add `onError` hook

### ISSUE THREAD

**Issue by delvedor** [fastify/fastify#1224]
## 🚀 Feature Proposal

Add a new application hook called `onError`, which a developer can use to track all the errors that Fastify is sending to the user.

## Motivation

Currently, we can't do that easily, because the `setErrorHandler` can be overwritten and the `onSend` hook receives the already serialized payload.
At the moment there is no way to get the error object and be sure at 100% that your code will never be overwritten.

Why do we need this?
https://github.com/elastic/apm-agent-nodejs/pull/594

Point of discussion:
- The `onError` hook should be detached from the request pipeline, in other words, it should not be able to influence the request/response (maybe allow just the headers)
- The user **cannot** call `reply.send` from the `onError` hook
- The hook should be async as the others or should behave as the `onRoute` hook?

## Example

```js
fastify.addHook('onError', (err, req, reply) => {
  // do something
})
```

I'm already working on a prototype, feedbacks are welcome!

**Comment by mcollina** [comment]
+1

**Comment by jsumners** [comment]
APM support would be welcome.

**Comment by delvedor** [comment]
Closed in #1236.
